### ISSUE TITLE
Saturation adjustment enhancements

### ISSUE THREAD

**Issue by charleskawczynski** [CliMA/Thermodynamics.jl#135]
## Purpose
@cmbengue requested an SDI be made to track several changes needed in Thermodynamics. This issue will outline them.

## Background
One concern we've had recently is whether our tolerance for saturation adjustment is reasonable or not because it's a bit difficult to reason about. If the tolerance is occasionally poor, we could have convergence issues with reasonable inputs and, at the moment, it's a bit difficult to know that since the inputs are not in more intuitive variables like temperature. #128 is an attempt to make sure that we start with a reasonable guess, however, knowing whether this is a good idea or not requires collecting data from a real-world run, and we don't have a clean way to do that at the moment. (msg copied from #132)

Some planned enhancements include:
 - Adding a data collection mechanism to help compare equations/formulations and solution method parameters (maximum number of iterations allowed, initial guesses, solution tolerance etc.)
 - Improve the accuracy, and hopefully convergence of the Clausius Clapeyron relation (#112)
 - Change tolerance to something more physically intuitive. There are two options here:
     - Change `ResidualTolerance` to `SolutionTolerance`
     - Change `e_tot(T) - e_tot_sat(T)` to `air_temperature(e_tot(T)) - air_temperature(e_tot_sat(T))` and still use `ResidualTolerance`. I tried this out locally and it does work (i.e., converges and passes tests)
 - One benefit I see with the latter is that we'll still be solving for `f(x) ≈ 0`, rather than solving for when `Δx_k < tol` for the `k`-th iteration.
 - Collect data from a real world run
     - Use the moist baro wave as an example
     - Run until clouds exist--when `(q_tot+q_ice) > threshold` (where `threshold ~ 0.01`? cc @tapios)

## Cost/Benefits/Risks
(Analyze the cost/benefits/risks associated with the proposal)

One question I have is: do we want to merge #132, or should we just use it during data collection and close the PR afterwards? We'll never want to use the data collection in production runs, I think. I don't have a strong preference either way.

I suppose that if we want to compare many different things, e.g., different `T_guess` estimates, different formulations, that require Thermodynamics code changes, it may be easier if it's part of the main branch. Maybe merging is just more practical

## People and Personnel
- Lead: @charleskawczynski
- Collaborators: @yairchn, @tapios
- Reviewers: @yairchn, @tapios

## Components
Components are in the mentioned PRs

## Inputs
Not applicable?

## Results and deliverables
Ideally, we'll be able to run moist baroclinic wave, with clouds, with 1 Newton Iteration and never hitting a non-converged example. We may need to tune the number of Newton iterations, but 

## Task Breakdown And Tentative Due Date
- [ ] #128
- [x] #132
- [x] #112
- [ ] #137
- [ ] #131
- [ ] #130
- [x] #32

## Proposed Delivery Date
Since this is relatively decoupled with other tasks, I've been working on this in parallel along other items. I'm thinking that it should be finished by end of next week, but I'm hoping on the earlier side.

**Comment by tapios** [comment]
This is a reasonable plan. A few comments:

1. A mechanism to collect average number of SA iterations is helpful. It would be nice as an optional tool in the main code. However, it's not strictly necessary, and relatively low priority. 

2. We should set the tolerance to a temperature tolerance. I'm fine with either way you propose.

3. If we cannot get real-world performance comparisons for the different initialization approaches in a relatively short time, the default should be to use the previous time step as initial value. This is how it's handled in other models. It's the tried and tested approach, to which we should default.
